src/bam/md_align.rs

```diff
@@ -0,0 +1,1152 @@
+    ///                       MatchDesc::Mismatch(b'G'),
+    ///                       MatchDesc::Matches(2),
+    ///                       MatchDesc::Matches(2),
+    ///                       MatchDesc::Deletion(b"AC".to_vec()),
+    ///                       MatchDesc::Mismatch(b'T'),
+    ///                       MatchDesc::Matches(5)];
+    /// let unnorm_str = MDString(unnorm_vec);
+    /// assert_eq!(unnorm_str.to_string(), "10A0G4^AC0T5");
+    /// # Ok( () )
+    /// # }
+    /// # fn main() { try_main().unwrap(); }
+    /// ```
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        let mut match_accum = 0;
+
+        let mut md_iter = self.0.iter().peekable();
+        while let Some(md) = md_iter.next() {
+            match md {
+                MatchDesc::Matches(mlen) => {
+                    if md_iter.peek().map_or(false, |next| next.is_matches()) {
+                        match_accum += mlen;
+                    } else {
+                        write!(f, "{}", match_accum + mlen)?;
+                        match_accum = 0;
+                    }
+                }
+                MatchDesc::Mismatch(refnt) => {
+                    if md_iter.peek().map_or(true, |next| !next.is_matches()) {
+                        write!(f, "{}0", *refnt as char)?;
+                    } else {
+                        write!(f, "{}", *refnt as char)?;
+                    }
+                }
+                MatchDesc::Deletion(refnts) => {
+                    if md_iter.peek().map_or(true, |next| !next.is_matches()) {
+                        write!(f, "^{}0", str::from_utf8(refnts).unwrap())?;
+                    } else {
+                        write!(f, "^{}", str::from_utf8(refnts).unwrap())?;
+                    }
+                }
+            };
+        }
+
+        Ok(())
+    }
+}
+
+impl<'a> IntoIterator for &'a MDString {
+    type Item = &'a MatchDesc;
+    type IntoIter = ::std::slice::Iter<'a, MatchDesc>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        (&self.0).into_iter()
+    }
+}
+
+impl IntoIterator for MDString {
+    type Item = MatchDesc;
+    type IntoIter = ::std::vec::IntoIter<MatchDesc>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+quick_error! {
+    #[derive(Debug,Clone)]
+    pub enum MDAlignError {
+        NoMD {
+            description("no MD aux field")
+        }
+        BadMD {
+            description("bad MD value")
+        }
+        MDvsCIGAR {
+            description("MD inconsistent with CIGAR")
+        }
+        BadSeqLen {
+            description("Sequence/quality length inconsistent with MD/CIGAR")
+        }
+        EmptyAlign {
+            description("Alignment has no positions")
+        }
+        ParseInt(err: ::std::num::ParseIntError) {
+            from()
+        }
+        Utf8(err: ::std::str::Utf8Error) {
+            from()
+        }
+    }
+}
```

Comment:
We've moved to thiseeror, hence, this should be adapted and moved into errors.rs.